core/rio/hdt/src/main/java/org/eclipse/rdf4j/rio/hdt/HDTParser.java

```diff
@@ -0,0 +1,311 @@
+			predicates = HDTDictionarySectionFactory.parse(bis);
+			predicates.parse(bis);
+
+			reportLocation(bis.getByteCount(), -1);
+			objects = HDTDictionarySectionFactory.parse(bis);
+			objects.parse(bis);
+
+			reportLocation(bis.getByteCount(), -1);
+			HDTTriples triples = new HDTTriples();
+			triples.parse(bis);
+
+			reportLocation(bis.getByteCount(), -1);
+			section = HDTTriplesSectionFactory.parse(new String(HDTTriples.FORMAT_BITMAP));
+			section.parse(bis);
+		} catch (IOException ioe) {
+			reportFatalError(ioe.getMessage(), bis.getCount(), -1);
+		} finally {
+			bis.close();
+		}
+
+		if (rdfHandler != null) {
+			rdfHandler.startRDF();
+		}
+
+		int cnt = 0;
+		int size = shared.size();
+
+		while (section.hasNext()) {
+			int[] t = section.next();
+			byte[] s = getSO(t[0], size, shared, subjects);
+			byte[] p = predicates.get(t[1]);
+			byte[] o = getSO(t[2], size, shared, objects);
+
+			Statement stmt = valueFactory.createStatement(createSubject(s), createPredicate(p), createObject(o));
+
+			if (rdfHandler != null) {
+				rdfHandler.handleStatement(stmt);
+			}
+		}
+		if (rdfHandler != null) {
+			rdfHandler.endRDF();
+		}
+	}
+
+	/**
+	 * Not supported, since HDT is a binary format.
+	 */
+	@Override
+	public synchronized void parse(Reader reader, String baseURI)
+			throws IOException, RDFParseException, RDFHandlerException {
+		throw new UnsupportedOperationException("HDT is binary, text readers not supported.");
+	}
+
+	/**
+	 * Get part of triple from shared HDT Dictionary or (if not found) from specific HDT Dictionary
+	 * 
+	 * @param pos    position
+	 * @param size   size of shared Dictionary
+	 * @param shared shared Dictionary
+	 * @param other  specific Dictionary
+	 * @return subject or object
+	 */
+	private byte[] getSO(int pos, int size, HDTDictionarySection shared, HDTDictionarySection other) {
+		return (pos <= size) ? shared.get(pos) : other.get(pos - size);
+	}
+
+	private boolean isBNodeID(byte[] b) {
+		// HDT-It generates "genid" for blank nodes in RDF/XML
+		return (b[0] == '_' || (b.length > 5 && b[0] == 'g' && b[1] == 'e'));
+	}
+
+	/**
+	 * Create subject IRI or blank node
+	 * 
+	 * @param b byte buffer
+	 * @return IRI or blank node
+	 */
+	private Resource createSubject(byte[] b) {
+		String str = new String(b, StandardCharsets.UTF_8);
+		return isBNodeID(b) ? valueFactory.createBNode(str) : valueFactory.createIRI(str);
+	}
+
+	/**
+	 * Create predicate IRI
+	 * 
+	 * @param b byte buffer
+	 * @return IRI
+	 */
+	private IRI createPredicate(byte[] b) {
+		return valueFactory.createIRI(new String(b, StandardCharsets.UTF_8));
+	}
+
+	/**
+	 * Create object (typed) literal, IRI or blank node
+	 * 
+	 * @param b byte buffer
+	 * @return literal, IRI or blank node
+	 */
+	private Value createObject(byte[] b) {
+		if (b[0] == '"') {
+			int i = b.length - 1;
+			for (; i > 1 && b[i] != '"'; i--) {
+				if (b[i] == '@') {
+					String lang = new String(b, i + 1, b.length - i - 1);
```

Comment:
The Charset version of this constructor should be used here (even if lang tags are always assumed to be simple-ASCII, we shouldn't leave these things to the users locale)